mesonbuild/backend/ninjabackend.py

```diff
@@ -321,6 +322,54 @@ def get_target_source_can_unity(self, target, source):
                 return False
         return True
 
+    def create_target_source_introspection(self, target: build.Target, comp: compilers.Compiler, parameters, sources, generated_sources):
+        '''
+        Adds the source file introspection information for a language of a target
+
+        Internal introspection storage formart:
+        self.introspection_data = {
+            '<target ID>': {
+                <id tuple>: {
+                    'language: 'lang',
+                    'compiler': ['comp', 'exe', 'list'],
+                    'parameters': ['UNIQUE', 'parameter', 'list'],
+                    'sources': [],
+                    'generated_sources': [],
+                }
+            }
+        }
+        '''
+        id = target.get_id()
+        lang = comp.get_language()
+        tgt = self.introspection_data[id]
+        # Find an existing entry or create a new one
+        id_hash = (lang, tuple(parameters))
+        src_block = tgt.get(id_hash, None)
+        if src_block is None:
+            # Convert parameters
+            if isinstance(parameters, CompilerArgs):
+                parameters = parameters.to_native(copy=True)
+            parameters = comp.compute_parameters_with_absolute_paths(parameters, self.build_dir)
+            if target.is_cross:
+                parameters.insert(0, comp.get_cross_extra_flags(self.environment, False))
```

Comment:
Maybe?
```suggestion
                parameters = comp.get_cross_extra_flags(self.environment, False) + parameters
```
insert(0, ...) will result in a nested list.

Not sure how this interacts with parameters still beeing a CompilerArgs here i think.